### DISCUSSION TITLE
Removing ble.sh from .bashrc prepends an @ to PS1

### DISCUSSION THREAD

**Discussion by Strykar** [akinomyoga/ble.sh#256] [Q&A] [answered]
If I comment out ble.sh from my .bashrc, my PS1 looks like `@[strykar@i-5 ~]$` instead of `[strykar@i-5 ~]$` since my `.bashrc` has `PS1='[\u@\h \W]\$ '`.
```
@[strykar@i-5 ~]$ echo $PS1
\[\e]133;k;start_kitty\a\]\[\e]133;A\a\]\[\e]133;k;end_kitty\a\][\u@\h \W]\$ \[\e]133;k;start_suffix_kitty\a\]\[\e[5 q\]\[\e]2;\w\a\]\[\e]133;k;end_suffix_kitty\a\]
@[strykar@i-5 ~]$ 

```
PS1 is fine after re-enabling ble.sh.
How do I troubleshoot this?

**Comment by akinomyoga** [comment]
> *From https://github.com/akinomyoga/ble.sh/issues/257*
> ```bash
> [strykar@i-5 ~]$ @echo $PS1
> [\u@\h \W]\$
>
> [strykar@i-5 ~]$ @
> ```
> If `PS1=[\u@\h \W]\$ ` the terminal should look like `[strykar@i-5 ~]$ ` instead of `[strykar@i-5 ~]$ @`.
> If I disable ble.sh, the prompt is prepended with the @ like @ [strykar@i-5 ~]$.

I am a bit confused. If it is the problem when you do not have ble.sh, how do you think it is related to ble.sh? Or, did you actually intend to write "If I enable ble.sh"? Hmm, or do you mean that an unexpected `@` is prepended without ble.sh and appended with ble.sh?

**Comment by Strykar** [comment]
The last, it is prepended without ble.sh and appended when ble.sh is sourced.

**Reply by akinomyoga** [reply]
OK, thank you for your clarification. If there is still a problem without ble.sh at all, currently, it doesn't seem to be an issue of ble.sh. But I guess you didn't know where to ask it so came here knowing that it might not be related to ble.sh?

**Reply by akinomyoga** [reply]
Let's see what is happening there. At least, I have never experienced such a phenomenon. What I would like to first confirm is if the phenomenon arises with another terminal. I guess from the above `PS1` that you are now using kitty? Is the same behavior observed with a different terminal emulator?

**Reply by Strykar** [reply]
> What I would like to first confirm is if the phenomenon arises with another terminal. I guess from the above `PS1` that you are now using kitty? Is the same behavior observed with a different terminal emulator?

Yes, it appears to be terminal agnostic.
Alacritty terminal:
```
[strykar@i-5 ~]$ @ echo "$PS1"                                                         ~
[\u@\h \W]\$
[strykar@i-5 ~]$ @
```

Gnome terminal:
```
[strykar@i-5 ~]$ @echo "$PS1"                                                  ~
[\u@\h \W]\$ 
[strykar@i-5 ~]$ @ 
```

Gnome Console:
```
[strykar@i-5 ~]$ @echo "$PS1"                                                  ~
[\u@\h \W]\$ 
[strykar@i-5 ~]$ @
```

**Reply by akinomyoga** [reply]
Thanks! Then, the next step is to check whether the problem reproduces with a plain Bash (with only a `PS1` setting in `.bashrc`). kitty terminal does something non-trivial even with an empty `~/.bashrc`, so I think it is more useful to use another terminal in order to identify the cause. Could you temporarily edit your bashrc and see if the situation changes with a simple .bashrc? Please do not forget to back up the original content of your bashrc. After checking the behavior you can recover the original settings from the backup (e.g. `~/.bashrc-backup` in the following example). For example,

```bash
$ cp -p ~/.bashrc ~/.bashrc-backup
$ echo "PS1='[\u@\h \W]\$ '" > ~/.bashrc
$ bash      # <-- start a new child session of bash
$           # <-- check how the prompt behaves
$ exit      # <-- after checking, you can end the child Bash session
$ cp ~/.bashrc-backup ~/.bashrc     # <-- restore the original content
```

**Reply by Strykar** [reply]
The [kitty author too thought](https://github.com/kovidgoyal/kitty/discussions/5729) it was my `.bashrc` but that was the first thing I looked at:
```
[strykar@i-5 ~]$ @cd                                                                                                                                                                                                                   ~
[strykar@i-5 ~]$ @cp -p ~/.bashrc ~/.bashrc-backup                                                                                                                                                                                     ~
[strykar@i-5 ~]$ @echo "PS1='[\u@\h \W]\$ '" > ~/.bashrc                                                                                                                                                                               ~
[strykar@i-5 ~]$ @bash                    # Note how the `@` gets prepended right after this                                                                                                                                                                                             ~
@[strykar@i-5 ~]$ exit
exit
[strykar@i-5 ~]$ @cp ~/.bashrc-backup ~/.bashrc                                                                                                                                                                                        ~
[strykar@i-5 ~]$ @bash                                                                                                                                                                                                                 ~

[strykar@i-5 ~]$ @exit                                                                                                                                                                                                                 ~
[ble: exit]

[strykar@i-5 ~]$ @
```

**Reply by akinomyoga** [reply]
OK, thank you for checking. Then what I suspect next is your `PROMPT_COMMAND` and `PS0`. Does the situation change after unsetting each of them?

```bash
$ unset PROMPT_COMMAND
$ unset PS0
```

**Reply by Strykar** [reply]
It does not change unfortunately:
```
[strykar@i-5 ~]$ @echo "$PROMPT_COMMAND"                                                                                                                                                                                               ~
printf "\033]0;%s@%s:%s\007" "${USER}" "${HOSTNAME%%.*}" "${PWD/#$HOME/\~}"
history -a; history -c; history -r

[strykar@i-5 ~]$ @echo "$PS0"                                                                                                                                                                                                          ~
\[\e]133;k;start_kitty\a\]\[\e]133;C\a\]\[\e]133;k;end_kitty\a\]\[\e]133;k;start_suffix_kitty\a\]\[\e[0 q\]$(_ksi_get_current_command)\[\e]133;k;end_suffix_kitty\a\]

[strykar@i-5 ~]$ @unset PROMPT_COMMAND                                                                                                                                                                                               ~
[strykar@i-5 ~]$ @unset PS0                                                                                                                                                                                                            ~
[strykar@i-5 ~]$ @                                                                                                                                                                                                                     ~
[strykar@i-5 ~]$ @
```

**Reply by akinomyoga** [reply]
OK, then could you try the Bash session started by the following command? This starts a Bash session in a completely clean environment and null settings. In the following command, please do not forget to specify a temporary filename to `HISTFILE`, or otherwise your command history `~/.bash_history` will be truncated.

```bash
$ env -i INPUTRC=/dev/null HOME=$HOME USER=$USER TERM=$TERM HISTFILE=~/a.tmp PS1='\w\$ ' bash --norc --noprofile
$               # <--- check if the situation changes
$ exit          # <--- close the session
$ rm -f ~/a.tmp # <--- remove the temporary file if it is created
```

**Reply by Strykar** [reply]
Progress! That does appear to move in the right direction:
```
[strykar@i-5 ~]$ @env -i INPUTRC=/dev/null HOME=$HOME USER=$USER TERM=$TERM HISTFILE=~/a.tmp PS1='\w\$ ' bash --norc --noprofile                                                  ~
~$ exit             # Note missing '@'
~$ exit
exit
[strykar@i-5 ~]$ @rm -f ~/a.tmp                                                                                                                                                   ~
[strykar@i-5 ~]$ @
```

**Reply by akinomyoga** [reply]
Great! Then, let's make the setting closer to the original one one by one.

First, let's remove `--noprofile`:

```bash
$ env -i INPUTRC=/dev/null HOME=$HOME USER=$USER TERM=$TERM HISTFILE=~/a.tmp PS1='\w\$ ' bash --norc
```

If the above does not reproduce the problem, what I suspect next is inputrc:

```bash
$ env -i HOME=$HOME USER=$USER TERM=$TERM HISTFILE=~/a.tmp PS1='\w\$ ' bash --norc
```

If the above does not reproduce the problem either, I think we also need to consider the content of `PS1`:

```bash
$ env -i HOME=$HOME USER=$USER TERM=$TERM HISTFILE=~/a.tmp PS1='[\u@\h \W]\$ ' bash --norc
```

**Reply by Strykar** [reply]
Only the first option appears to remedy the situation:
```
[strykar@i-5 ~]$ @env -i INPUTRC=/dev/null HOME=$HOME USER=$USER TERM=$TERM HISTFILE=~/a.tmp PS1='\w\$ ' bash --norc                                                                                                                                                                       ~
~$ exit            # Notice the missing '@' here
exit
[strykar@i-5 ~]$ @env -i HOME=$HOME USER=$USER TERM=$TERM HISTFILE=~/a.tmp PS1='\w\$ ' bash --norc                                                                                                                                                                                         ~
@~$ exit
exit
[strykar@i-5 ~]$ @env -i HOME=$HOME USER=$USER TERM=$TERM HISTFILE=~/a.tmp PS1='[\u@\h \W]\$ ' bash --norc                                                                                                                                                                                 ~
@[strykar@i-5 ~]$
```

**Reply by akinomyoga** [reply]
<details><summary>(Please forget about this comment. I think this is irrelevant now.)</summary>
OK! If so, I guess the problem does not appear with the session started as the following. Is it correct?

```bash
$ bash --noprofile
$      # <-- check the behavior
$ exit # <-- close the session
```
</details>

**Reply by akinomyoga** [reply]
This means that your `~/.inputrc` is related to the behavior.

Ah, OK. I think now I understand what is happening. Don't you have the following line in your `~/.inputrc`?

```inputrc
set show-mode-in-prompt on
```

**Reply by Strykar** [reply]
Fixed!! Thank you so much!
I had added that years ago and forgotten about it:
```
#set editing-mode vi
set show-mode-in-prompt on
```

**Reply by akinomyoga** [reply]
Great. Let me summarize the solution in another reply.

**Comment by akinomyoga** [comment, marked as answer]
Here's the summary. This turned out to be caused by the readline setting `set show-mode-in-prompt on`. This is a feature of GNU Readline, and the character `@` is used to represent that we are currently in Emacs editing mode. You can remove or comment out the corresponding line in your `~/.inputrc`, `~/.bashrc`, etc.

When `~/.inputrc` contains the setting:

```inputrc
# inputrc

# If you have the following line, you can comment it out or remove it
set show-mode-in-prompt on
```

When `~/.bashrc` or other files sourced from `~/.bashrc` contains the setting:

```bash
# bashrc

# If you have the following line in bashrc, you can comment it out or remove it
bind 'set show-mode-in-prompt on'
```

**Comment by Strykar** [comment]
Thank you for taking the time to help me troubleshoot this, and for your amazing software `ble.sh`, whose surface and power, I have barely scratched! :pray:
